R/pipeline.R

```diff
@@ -0,0 +1,221 @@
+#' other pipeline artifacts will be saved. Defaults to the root directory ("/").
+#'
+#' @details
+#' The function reads the configuration from a JSON file and uses this to set
+#' up the job and task identifiers. It creates an output directory structure
+#' based on these IDs and starts logging the process in a file. The main
+#' pipeline process is handled by `process_pipeline()`, with errors caught and
+#' logged as warnings. The function will log the success or failure of the run.
+#'
+#' Logs are written to a file in the output directory, and console output is
+#' also mirrored in this log file. Error handling is in place to capture any
+#' issues during the pipeline execution and ensure they are logged
+#' appropriately.
+#'
+#' During the execution of the pipeline, the following output files are
+#' expected to be generated:
+#'
+#' - **Model Output**: An RDS file of the fitted model is saved in the
+#' task-specific directory (`model.rds`).
+#' - **Samples**: Parquet files containing the model's sample outputs are saved
+#' in a `samples` subdirectory, named using the `task_id` (e.g.,
+#' `task_id.parquet`).
+#' - **Summaries**: Parquet files summarizing the model's results are saved in
+#' a `summaries` subdirectory, also named using the `task_id` (e.g.,
+#' `task_id.parquet`).
+#' - **Logs**: A `logs.txt` file is generated in the task directory, capturing
+#' both console and error messages.
+#'
+#' The output directory structure will follow this format:
+#' ```
+#' <output_dir>/
+#' └── <job_id>/
+#'     ├── samples/
+#'     │   └── <task_id>.parquet
+#'     ├── summaries/
+#'     │   └── <task_id>.parquet
+#'     └── tasks/
+#'         └── <task_id>/
+#'             ├── model.rds
+#'             └── logs.txt
+#' ```
+#'
+#' @return The function does not return a value directly. However, the output
+#' directory will contain the following files:
+#' - Model RDS file (`model.rds`)
+#' - Sample output in Parquet format (`<task_id>.parquet` in the `samples/`
+#' directory)
+#' - Summary output in Parquet format (`<task_id>.parquet` in the `summaries/`
+#' directory)
+#' - Log file (`logs.txt`) in the task directory
+#'
+#' @rdname pipeline
+#' @export
+run_pipeline <- function(config_path,
+                         blob_storage_container = NULL,
+                         output_dir = "/") {
+  # TODO: Add config reader here
+  config <- jsonlite::read_json(config_path,
+    simplifyVector = TRUE
+  )
+
+  write_output_dir_structure(
+    output_dir = output_dir,
+    job_id = config[["job_id"]],
+    task_id = config[["task_id"]]
+  )
+
+  # Set up logs
+  logfile_path <- file.path(
+    output_dir,
+    config[["job_id"]],
+    "tasks",
+    config[["task_id"]]
+  )
+  logfile_connection <- file(file.path(logfile_path, "logs.txt"), open = "wt")
+  sink(
+    logfile_connection,
+    type = "output",
+    append = TRUE,
+    # Send output to logs and to console
+    split = TRUE
+  )
```

Comment:
Running this locally with Docker, I don't see output to the console. If I enter the container, and run the pipeline from there, I still don't see output printed to the console, but I can see the output log file and read that.

Reply:
Ah yes, I need to take that out because it doesn't actually work as documented. I'll turn it into an issue.

Reply:
Ok, so I just ran again locally this morning, and this time it actually did print to the console. I'm confused.